src/serac/physics/utilities/variational_form/integral.hpp

```diff
@@ -0,0 +1,591 @@
+    // into the quadrature function in the finite element kernel.
+    //
+    // we use them to observe the output type and allocate memory to store
+    // the derivative information at each quadrature point
+    using x_t             = tensor<double, spatial_dim>;
+    using u_du_t          = typename lambda_argument<trial_space, geometry_dim, spatial_dim>::type;
+    using derivative_type = decltype(get_gradient(qf(x_t{}, make_dual(u_du_t{}))));
+
+    auto num_quadrature_points = static_cast<uint32_t>(X.Size() / spatial_dim);
+    qf_derivatives_.resize(sizeof(derivative_type) * num_quadrature_points);
+
+    // FIXME: Strict aliasing rule?
+    auto qf_derivatives_ptr = reinterpret_cast<derivative_type*>(qf_derivatives_.data());
+
+    // this is where we actually specialize the finite element kernel templates with
+    // our specific requirements (element type, test/trial spaces, quadrature rule, q-function, etc).
+    //
+    // std::function's type erasure lets us wrap those specific details inside a function with known signature
+    //
+    // note: the qf_derivatives_ptr is copied by value to each lambda function below,
+    //       to allow the evaluation kernel to pass derivative values to the gradient kernel
+    evaluation_ = [=](const mfem::Vector& U, mfem::Vector& R) {
+      evaluation_kernel<geometry, test_space, trial_space, geometry_dim, spatial_dim, Q>(U, R, qf_derivatives_ptr, J_,
+                                                                                         X_, num_elements, qf);
+    };
+
+    gradient_ = [=](const mfem::Vector& dU, mfem::Vector& dR) {
+      gradient_kernel<geometry, test_space, trial_space, geometry_dim, spatial_dim, Q>(dU, dR, qf_derivatives_ptr, J_,
+                                                                                       num_elements);
+    };
+  }
+
+  /**
+   * @brief Applies the integral, i.e., @a output_E = evaluate( @a input_E )
+   * @param[in] input_E The input to the evaluation; per-element DOF values
+   * @param[out] output_E The output of the evalution; per-element DOF residuals
+   * @see evaluation_kernel
+   */
+  void Mult(const mfem::Vector& input_E, mfem::Vector& output_E) const { evaluation_(input_E, output_E); }
+
+  /**
+   * @brief Applies the integral, i.e., @a output_E = gradient( @a input_E )
+   * @param[in] input_E The input to the evaluation; per-element DOF values
+   * @param[out] output_E The output of the evalution; per-element DOF residuals
+   * @see gradient_kernel
+   */
+  void GradientMult(const mfem::Vector& input_E, mfem::Vector& output_E) const { gradient_(input_E, output_E); }
+
+private:
+  /**
+   * @brief Jacobians of the element transformations at all quadrature points
+   */
+  const mfem::Vector J_;
+  /**
+   * @brief Mapped (physical) coordinates of all quadrature points
+   */
+  const mfem::Vector X_;
+
+  /**
+   * @brief Byte array of derivative data
```

Comment:
I'm not sure if this is the right place, but it would be good to document the striding/layout of the derivatives w.r.t to component and shape-function/elem_dof.

Reply:
fixed in another branch